prow/plugins/projectmanager/projectmanager.go

```diff
@@ -0,0 +1,204 @@
+	pluginHelp := &pluginhelp.PluginHelp{
+		Description: "The project-manager plugin automatically adds Pull Requests to specified GitHub Project Columns if they match given criteria.",
+		Config: func(config *plugins.Configuration) map[string]string {
+			configMap := make(map[string]string)
+			configString := "org/repos: {"
+			for orgRepoName, managedOrgRepo := range config.ProjectManager.OrgRepos {
+				configString := fmt.Sprintf("%s %s: { projects: {", configString, orgRepoName)
+				for projectName, managedProject := range managedOrgRepo.Projects {
+					configString := fmt.Sprintf("%s %s: { columns: [", configString, projectName)
+					for _, managedColumn := range managedProject.Columns {
+						configString := fmt.Sprintf("%s {id: \"%d\", name: \"%s\", state: \"%s\", org: \"%s\" labels: [", configString, managedColumn.ID, managedColumn.Name, managedColumn.State, managedColumn.Org)
+						for i, label := range managedColumn.Labels {
+							configString = fmt.Sprintf("%s \"%s\"", configString, label)
+							if i+1 < len(managedColumn.Labels) {
+								configString = fmt.Sprintf("%s,", configString)
+							}
+						}
+						configString = fmt.Sprintf("%s ] }", configString)
+					}
+					configString = fmt.Sprintf("%s ] }", configString)
+				}
+				configString = fmt.Sprintf("%s }", configString)
+			}
+			configString = fmt.Sprintf("%s }", configString)
+			configMap[""] = configString
+			return configMap
+		}(config),
+	}
+	return pluginHelp, nil
+}
+
+func handlePullRequest(pc plugins.Agent, pe github.PullRequestEvent) error {
+	return handlePR(pc.GitHubClient, pc.PluginConfig.ProjectManager, pc.Logger, pe)
+}
+
+// Strict subset of *github.Client methods.
+type githubClient interface {
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+	GetRepoProjects(owner, repo string) ([]github.Project, error)
+	GetOrgProjects(org string) ([]github.Project, error)
+	GetProjectColumns(projectID int) ([]github.ProjectColumn, error)
+	CreateProjectCard(columnID int, projectCard github.ProjectCard) (*github.ProjectCard, error)
+}
+
+func handlePR(gc githubClient, projectManager plugins.ProjectManager, log *logrus.Entry, pe github.PullRequestEvent) error {
+	// Only respond to label add or issue/PR open events
+	if pe.Action != github.PullRequestActionOpened &&
+		pe.Action != github.PullRequestActionReopened &&
+		pe.Action != github.PullRequestActionLabeled {
+		return nil
+	}
+	// Get any ManagedProjects that match this PR
+	matchedColumnIDs, err := getMatchingColumnIDs(gc, projectManager, pe)
+	if err != nil {
+		return err
+	}
+	// For each ManagedColumn that matches this PR, add this PR to that Project Column
+	for _, matchedColumnID := range matchedColumnIDs {
+		err = addPRToColumn(gc, matchedColumnID, pe)
+		log.WithError(err).Println("Failed to add PR to project")
+	}
+	return nil
+}
+
+func getMatchingColumnIDs(gc githubClient, projectManager plugins.ProjectManager, pe github.PullRequestEvent) ([]int, error) {
```

Comment:
- Looks like we are only using `projectManager.OrgRepos` in this method, so we can pass that in instead of the entire `plugins.ProjectManager`
- We can call `gc.GetIssueLabels(...)` in `handlePR(...)` instead of calling `gc.GetIssueLabels(...)` in the nested for loops.

Reply:
@taragu Not true, projectManager has a set of Managed Projects and a set of ManagedColumns embedded, we need to pass the entire structure to validate if the PR meets the criteria.

Reply:
`projectManager has a set of Managed Projects and a set of ManagedColumns embedded` --> @brahmaroutu  that's right. Managed repos, projects, columns are all under `projectManager.OrgRepos`. So can we pass `projectManager.OrgRepos` into this function instead of `projectManager`?

Reply:
Done

Reply:
I am not sure if there is any advantage to passing in  map[string]ManagedOrgRepo vs ProjectManager object?

Reply:
@brahmaroutu I suggested this for better readability. If we pass in `map[string]ManagedOrgRepo` then readers immediately know from the function signature that this is the piece of data being used, versus if we pass in `plugins.ProjectManager`, readers will need to read the content of this function to see that only `projectManager.OrgRepos` is used. This is minor and you can leave this as is. It's completely up to you if you would like to make this change.